Definition infrastructure for instruction count ubenchmarks

robieta opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53296 Definition infrastructure for instruction count ubenchmarks**
* #54484 Revert "Revert D26907093: Add repeats to Timer.collect_callgrind(...)"

Part 1 of the instruction count microbenchmarks. This PR is focused on benchmark definition machinery. (Though you can run `main.py` to see it in action.) A summary of the system is given in the README.

Differential Revision: [D26907092](https://our.internmc.facebook.com/intern/diff/D26907092)

Commit:
Definition infrastructure for instruction count ubenchmarks

[ghstack-poisoned]

Commit:
Update on "Definition infrastructure for instruction count ubenchmarks"




[ghstack-poisoned]

Commit:
Update on "Definition infrastructure for instruction count ubenchmarks"




[ghstack-poisoned]

Commit:
Update on "Definition infrastructure for instruction count ubenchmarks"




[ghstack-poisoned]

Commit:
Update on "Definition infrastructure for instruction count ubenchmarks"




[ghstack-poisoned]

Commit:
Update on "Definition infrastructure for instruction count ubenchmarks"


Part 1 of the instruction count microbenchmarks. This PR is focused on benchmark definition machinery. (Though you can run `main.py` to see it in action.) A summary of the system is given in the README.

[ghstack-poisoned]

ezyang commented on benchmarks/instruction_counts/README.md:
```diff
@@ -0,0 +1,133 @@
+(managing the model artifact, setting up IValues) which if done manually would
+be rather bug-prone and hard to read.
+
+The story is similar for AutoGrad: because we know the output variable (`y`)
```
nit: it's autograd, not AutoGrad

robieta replied:
Done.

ezyang commented on benchmarks/instruction_counts/definitions/standard.py:
```diff
@@ -0,0 +1,72 @@
+    "Empty": {
+        "no allocation": GroupedStmts(
+            r"torch.empty(())",
+            r"torch::empty({0});",
```
I'm wondering how this format will scale if you want to do really detailed, feature-by-feature instruction count benchmarking. For example, a while ago @swolchok was remarking how hitting the TensorOptions overload in empty was worse than the faithful overload, even when you were passing in a literal (e.g., `torch::empty({0}, at::kDouble)`) because C++ had to pack the literal into a TensorOptions, and then immediately unpack it to do the dispatch. The benchmark here isn't going to capture that; if you want to accurately capture the overheads of all the ways to call torch::empty, you're probably going to need a dozen or so variations (`torch::empty({0}, options)`, `torch::empty({0}, dtype)`, `torch::empty({0}, scalartype)`, etc). Looking at the current format, it seems that for every single line (or two, I guess, with the Python variant) of benchmark code, I need to write four more lines of boilerplate; that's more than doubling the size of the test in question. I think I, personally, would appreciate a much shorter format, esp, if I am going to have to do a lot of different variants.

ezyang replied:
Based on the README, it also sounds like one of the things a lot of investment was made into was making it possible to run a benchmark under as many variations of the frontend as possible (so, C++, Python, TorchScript, etc.), but if I'm like painstakingly enumerating a lot of different torch.empty variations because I happen to know that these variations matter due to C++ implementation details, that doesn't necessarily mean that the TorchScript interpreter's performance faults cut along the same lines. So it might actually just be a waste of time to run each variation under TS, because actually you need a different set of variations to exercise different paths in the interpreter.

For microbenchmarks like this, I think it is legitimate to treating the system as a white box when deciding what to cover, because the point of instruction counts are to be actionable, and so when a benchmark fails we want this to correlate to some specific portion of code we ought to turn our attention to.

robieta replied:
Absolutely. Terse clear benchmark definitions are VERY important. I think generally the way as it currently stands there are two ways. So for instance in the example above:

`Option A: TimerArgs per line`

```
"Empty": {
    "no allocation": GroupedStmts(
        r"torch.empty(())",
        r"torch::empty({0});",
    ),
    "literal": TimerArgs("torch::empty({0}, at::kDouble);", language="cpp"),
    "TensorOptions": TimerArgs("torch::empty({0}, options);", "auto options = ...", language="cpp"),
    "ScalarType: TimerArgs("torch::empty({0}, scalartype);", "auto scalartype = ...", language="cpp"),
    ...
}
```

Pros:
* One line per def
* WYSIWYG
  
Cons:
* Fluff like `TimerArgs` and `language="cpp"` on each line.
* Nothing is aligned which makes it hard to read.

`Option B: Use comprehensions`

```
# This is a limitation of dicts. Without the `None` the later use of "Empty" would blow away this one and there's no way
# to detect and preserve it.
("Empty", None): {
    "no allocation": GroupedStmts(
        r"torch.empty(())",
        r"torch::empty({0});",
    ),
},
"Empty": {
    label: TimerArgs(f"torch::empty({{0}}, {args});", setup=setup, language="cpp")
    for label, args, setup in (
        ("literal", "at::kDouble", ""),
        ("TensorOptions", "options", "auto options = ..."),
        ("ScalarType", "scalartype", "auto scalartype = ..."),
        ...
    )
},
```

Pros:
* Scales very well. Signal to noise in the "definition" portion is nearly 100%
* Any sort of repeated code gets amortized out.
  
Cons:
* There is an O(5-10) line count overhead
* It will get very hard to read if one gets too clever. (True of all comprehensions, really.)

I'm imagining (A) would be for cases with 2-4 variants where you just don't want to pay the newline padding cost you describe, and (B) would be for where you have LOTS of cases but they're mechanical, and you can just read the comprehension args and get a very terse description of what is going on.

In the bigger set of benchmarks, I wound up using this sort of trick for indexing (https://github.com/pytorch/pytorch/pull/50608/files#diff-e4698f454a1a411efa34e153cedfe21dd06a129527fce585bfff813bdaa8455bR127-R152) and nn Modules (https://github.com/pytorch/pytorch/pull/50608/files#diff-e4698f454a1a411efa34e153cedfe21dd06a129527fce585bfff813bdaa8455bR224-R250) if you want to use that to gauge the "comprehension" approach to readable similar benchmarks.

But of course if you can come up with a more readable format I'm all ears.

robieta replied:
I saw the second comment after I wrote my response, but in summary: `TimerArgs` is the white box variant.

ngimel replied:
I think the question is how to write things similar to what we have in tests:
```
for kwarg1, kwarg2, kwarg3 in product((True, False), (0,1,2), (torch.half, torch.float)):
     f(kwarg1=kwarg1, kwarg2=kwarg2, kwarg3=kwarg3)
```
I feel there must be a way to write it with a comprehension, the question is how readable and writeable it is

bhosmer replied:
@robieta just to clarify, in the A/B examples you gave above, the "no allocation" tests are just the existing benchmarks, correct? i.e. if I'm just doing something local I can drop them. and it's just a choice between e.g.
```
{
    "literal": TimerArgs("torch::empty({0}, at::kDouble);", language="cpp"),
    "TensorOptions": TimerArgs("torch::empty({0}, options);", "auto options = ...", language="cpp"),
    "ScalarType: TimerArgs("torch::empty({0}, scalartype);", "auto scalartype = ...", language="cpp"),
    ...
}
```
and
```
{
    label: TimerArgs(f"torch::empty({{0}}, {args});", setup=setup, language="cpp")
    for label, args, setup in (
        ("literal", "at::kDouble", ""),
        ("TensorOptions", "options", "auto options = ..."),
        ("ScalarType", "scalartype", "auto scalartype = ..."),
        ...
    )
}
```
If so, then how far from reality is this transcription of Natalia's example? Judging by the cardinality of `label` in your example it doesn't seem like comprehensions have the same combinatorial behavior as `product` - if that's accurate, would it be feasible to add? (We'd need some way to generate the necessary `label` entries ofc but I'm not sure what else would be needed)
```
TimerArgs(f"f(kwarg1={kwarg1}, kwarg2={kwarg2}, kwarg3={kwarg3});", 
    for kwarg1, kwarg2, kwarg3 in ((true, false), (0, 1, 2), (torch.half, torch.float))
)
```

ezyang replied:
The emphasis on list comprehensions still feels a little off to me.

Let me describe my thought process for how I would setup microbenchmarks for, say torch::empty.

For torch::empty, I'm primarily interested in dispatcher overhead, and exercising the different overloads. So when writing this benchmark I take a look at the empty overloads (ignoring named tensors for the moment):

```
TORCH_API Tensor empty(IntArrayRef size, TensorOptions options={}, c10::optional<MemoryFormat> memory_format=c10::nullopt);
TORCH_API Tensor empty(IntArrayRef size, c10::optional<ScalarType> dtype, c10::optional<Layout> layout, c10::optional<Device> device, c10::optional<bool> pin_memory, c10::optional<MemoryFormat> memory_format);
```

I want to exercise each of these overloads. So I sit down and start writing down calls which express the different overloads:

```
int main() {

  // Setup
  auto options_empty = c10::TensorOptions();
  auto options_full = c10::TensorOptions().dtype(kFloat).device(kCPU);
  auto optional_float = c10::make_optional(kFloat);

  // TensorOptions overload
  at::empty({0}, options_empty);
  at::empty({0}, options_full);
  at::empty({0}, at::kFloat); // implicit conversion

  // Faithful overload
  at::empty({0}, nullopt, nullopt, nullopt, nullopt, nullopt);
  at::empty({0}, kFloat, nullopt, nullopt, nullopt, nullopt);
  at::empty({0}, at::optional_float, nullopt, nullopt, nullopt, nullopt);

  // Baseline comparison
  at::empty({0});
  at::cpu::empty({0});

  return 0;
}
```

I like that it's a little C++ program, since that means I can paste it into Godbolt and look at the assembly easily. Because I'm whiteboxing, I'm not interested in the full combinatorial explosion of possibilities (e.g., I don't test everything on both at::cpu and at::), because I happen to know that the relevant TensorOptions unpacking code is shared between the two implementations, so one comparison is good enough. Because I'm specifically interested in the overload handling, I purposely don't exercise differing sizes; that's benchmarking something else, and maybe I have a separate test program for that.

Now, I need to translate this into benchmarks inside the framework. Honestly, this program (maybe with some small modifications, e.g., no int main wrapper) ought to be enough:

1. I don't want to have to name my individual benchmarks. Each line of code is super self explanatory, and should be translated into the name itself. Maybe I give an explanatory comment for one or subtle cases.
2. I don't want to have to manage propagating the setup lines to the appropriate benchmark that uses it. I define one setup block and use it for all the benchmarks. Maybe if the setup is time consuming and I don't need it for some other benchmark, I DCE the useless setup lines.
3. I don't want to have to setup a list comprehension. One line, one benchmark.

ezyang replied:
I think what it is that I really hate about list comprehensions is, from the reader's perspective, it obscures what the *actual* C++ code that is being run under the benchmark is. I don't mind abstracting away setup code, but for the actual thing I'm measuring, I really *do* want to see the code in question, I don't want it DRYed away

robieta replied:
@ezyang @bhosmer @ngimel I've added a `GroupedVariants` definition scheme, which basically implements the "benchmark-per-line" approach discussed earlier. It also seems to compose reasonably well with `parse_stmts`; for instance indexing benchmarks (in the other stack) weren't particularly ergonomic with comprehensions, whereas now it's much clearer what's going on.

Commit:
Update on "Definition infrastructure for instruction count ubenchmarks"


Part 1 of the instruction count microbenchmarks. This PR is focused on benchmark definition machinery. (Though you can run `main.py` to see it in action.) A summary of the system is given in the README.

[ghstack-poisoned]

ngimel commented on benchmarks/instruction_counts/definitions/setup.py:
```diff
@@ -0,0 +1,57 @@
+"""Define some common setup blocks which benchmarks can reuse."""
```
It would make sense to put parser limitations in this file, many people will look at the examples here, but will never know the rules to obey

robieta replied:
Done.

ngimel reviewed (commented):
I've started going over this PR. It's much easier to review and understand what's going on in this format! 
General structure looks good to me.

Commit:
Update on "Definition infrastructure for instruction count ubenchmarks"


Part 1 of the instruction count microbenchmarks. This PR is focused on benchmark definition machinery. (Though you can run `main.py` to see it in action.) A summary of the system is given in the README.

Differential Revision: [D26907092](https://our.internmc.facebook.com/intern/diff/D26907092)

[ghstack-poisoned]

Commit:
Update on "Definition infrastructure for instruction count ubenchmarks"


Part 1 of the instruction count microbenchmarks. This PR is focused on benchmark definition machinery. (Though you can run `main.py` to see it in action.) A summary of the system is given in the README.

Differential Revision: [D26907092](https://our.internmc.facebook.com/intern/diff/D26907092)

[ghstack-poisoned]

robieta commented on benchmarks/instruction_counts/README.md:
````diff
@@ -0,0 +1,133 @@
+```
+
+Moreover, because `signature` is provided we know that creation of `x` and `w`
+is part of setup, and the overall comptation uses `x` and `v` to produce `y`.
````
Done.

robieta commented on benchmarks/instruction_counts/core/api.py:
```diff
@@ -0,0 +1,328 @@
+
+@dataclasses.dataclass(frozen=True)
+class GroupedSetup:
+    py_setup: Optional[str] = None
```
Indeed. There are some cases where `None` literally means "this benchmark was not defined", but for the cases where I was just overusing `Optional` I've replaced them with just `str`.

robieta commented on benchmarks/instruction_counts/README.md:
````diff
@@ -0,0 +1,133 @@
+```
+# `GroupedStmts` is an alias of `GroupedBenchmark.init_from_stmts`
+benchmark = GroupedStmts(
+    py_stmt=r"y = x * w",
````
No. `py_fwd_stmt` and `cpp_fwd_stmt` are names of fields, but the construction methods take `py_stmt` and `cpp_stmt`. I thought having `fwd` in the interface might be confusing.

robieta commented on benchmarks/instruction_counts/README.md:
````diff
@@ -0,0 +1,133 @@
+python main.py
+```
+
+Currently `main.py` contains a very simple threadpool (so that run time isn't
````
Yeah, it was busted. (Part of the great `Optional` cull, but I didn't get all the sources of None's.) Should be fixed now. Sorry about that.

ngimel reviewed (approved):
This looks good to me. Late addition of GroupedVariants is a bit undocumented, both it its purpose and in parsing limitations (e.g. space after `<--` matters, `<--` separators can be used to label following statements etc), but functionality itself is good, and I agree indexing definitions with it are more readable. 
Great work!

bhosmer commented on benchmarks/instruction_counts/definitions/standard.py:
```diff
@@ -0,0 +1,131 @@
+
+        "overloads": GroupedVariants(
+            cpp_block=r"""
+                // <-- Setup -->
```
Total bikeshed, but fwiw the `<--` `-->` seems a little heavy to me, if all we want to do is distinguish the label comments from ordinary ones. I'd be inclined to pitch just like `// @label` - but it's a minor point. Totally your call.

Regardless though, I agree w/Natalia that `GroupedVariants` is a little underdescribed considering how significantly it increases the expressiveness of the system. Given that it's not the job of this code to provide complete documentation, what about just 
* a line early in the README that points to this file for examples, and 
* a comment header at the top of this file that gives for constraints for each format (i.e., roughly what you have at the top of the `parse_stmts` function)

Not to do the work of an actual tutorial, just so if you're landing here, you have enough to generalize from these examples with a minimum of trial and error.

bhosmer reviewed (approved):
Excellent! And really nice work on `GroupedVariants` - basically exactly the idiom requested by your focus group 😁  couple minor ideas inline (mostly just about docs/discoverability), but yeah, this is an awesome toolkit.

Commit:
Update on "Definition infrastructure for instruction count ubenchmarks"


Part 1 of the instruction count microbenchmarks. This PR is focused on benchmark definition machinery. (Though you can run `main.py` to see it in action.) A summary of the system is given in the README.

Differential Revision: [D26907092](https://our.internmc.facebook.com/intern/diff/D26907092)

[ghstack-poisoned]

robieta commented on benchmarks/instruction_counts/core/api.py:
```diff
@@ -0,0 +1,398 @@
+
+    Concrete interfaces:
+     - `core.api.GroupedStmts`     (init_from_stmts)
+     - `core.api.GroupedModules`   (init_from_model)
```
Good point. Added.

Commit:
Update on "Definition infrastructure for instruction count ubenchmarks"


Part 1 of the instruction count microbenchmarks. This PR is focused on benchmark definition machinery. (Though you can run `main.py` to see it in action.) A summary of the system is given in the README.

Differential Revision: [D26907092](https://our.internmc.facebook.com/intern/diff/D26907092)

[ghstack-poisoned]

robieta commented:
@ngimel @bhosmer 

I made the following doc tweaks:
* Added a section to the README quick start section with a one line summary of each of the ways to define a benchmark. (Give people a starting point to start grep-ing)
* Expanded the docstring in `api.py`, and moved the parser details summary from `setup.py` to `standard.py`. (Adding entries for GroupedVariants.)
* Changed the label syntax to Basil's `@label` syntax. (I agree that it is simpler, and it addresses Natalia's concern about documenting whitespace by making it a moot point.)

ngimel commented:
Great, thank you!

Commit:
Update on "Definition infrastructure for instruction count ubenchmarks"


Part 1 of the instruction count microbenchmarks. This PR is focused on benchmark definition machinery. (Though you can run `main.py` to see it in action.) A summary of the system is given in the README.

Differential Revision: [D26907092](https://our.internmc.facebook.com/intern/diff/D26907092)

[ghstack-poisoned]

Commit:
Update on "Definition infrastructure for instruction count ubenchmarks"


Part 1 of the instruction count microbenchmarks. This PR is focused on benchmark definition machinery. (Though you can run `main.py` to see it in action.) A summary of the system is given in the README.

Differential Revision: [D26907092](https://our.internmc.facebook.com/intern/diff/D26907092)

[ghstack-poisoned]
